src/utilities/axisBumpFactory.ts

```diff
@@ -0,0 +1,68 @@
+// Map<YAxis, Map<NodeId, offset>>
+type AxisBumps = Map<number, Map<string, number> | undefined>
+
+type SetBumpParameters = {
+  axis: number,
+  nodeId: string,
+  offset: number,
+}
+
+type RemoveBumpParameters = {
+  axis: number,
+  nodeId: string,
+}
+
+type BumpFactory = {
+  get: (axis: number) => number,
+  set: (value: SetBumpParameters) => void,
+  delete: (value: RemoveBumpParameters) => void,
+  bump: (axis: number) => number,
+  clear: () => void,
+}
+
+export function axisBumpFactory(): BumpFactory {
+  const bumps: AxisBumps = new Map()
+
+  const get: BumpFactory['get'] = (axis) => {
+    const values = bumps.get(axis)
+
+    if (!values) {
+      return 0
+    }
+
+    return Math.max(...values.values())
+  }
+
+  const set: BumpFactory['set'] = ({ axis, nodeId, offset }) => {
+    const value = bumps.get(axis) ?? new Map<string, number>()
+
+    value.set(nodeId, offset)
+    bumps.set(axis, value)
+  }
+
+  const remove: BumpFactory['delete'] = ({ axis, nodeId }) => {
+    bumps.get(axis)?.delete(nodeId)
+  }
+
+  const bump: BumpFactory['bump'] = (axis) => {
```

Comment:
This verb reads more like a set than a get to me? Maybe it can just be called `get`

Reply:
Agreed. I ended up naming this `getTotalOffset` with calls `getOffset`. 

- `getOffset` - just that rows offset
- `getTotalOffset` - the total offset factoring in previous rows